### ISSUE TITLE
Getting NPE on AsyncSubject.unsubscribe

### ISSUE THREAD

**Issue by YulawOne** [monix/monix#308]
This cause stream hangs, and after it may never complete.
Below is minimal code sample to reproduce error:
```scala
  implicit val scheduler = Scheduler(ExecutionContext.global)

  val ticks = Observable.interval(1.second).map(_ => ()).whileBusyDropEvents

  val loop =
    ticks
      .mapTask { _ =>
        {
          for {
            random <- Task.eval(Random.nextInt(10))
            result <-
              if (random < 6) Task.pure(Some(s"Hi $random"))
              else if (random >= 9) Task.raiseError(new RuntimeException("ops"))
              else Task.pure(None)
          } yield result
        }
         .restartUntil(_.isEmpty)
         .map(_ => Done)
      }

  val active = loop.multicast(Pipe.async[Done]).refCount

  val switch = ticks.switchMap(_ => active).completedL

  Await.result(switch.runAsync, Duration.Inf)
```

```
java.lang.NullPointerException
	at monix.reactive.subjects.AsyncSubject.unsubscribe(AsyncSubject.scala:122)
	at monix.reactive.subjects.AsyncSubject.$anonfun$unsafeSubscribeFn$1(AsyncSubject.scala:89)
	at monix.execution.Cancelable$CancelableTask.cancel(Cancelable.scala:66)
	at monix.reactive.observables.RefCountObservable.$anonfun$unsafeSubscribeFn$2(RefCountObservable.scala:65)
	at monix.execution.Cancelable$CancelableTask.cancel(Cancelable.scala:66)
	at monix.execution.cancelables.SerialCancelable.cancel(SerialCancelable.scala:62)
	at monix.execution.cancelables.CompositeCancelable.cancel(CompositeCancelable.scala:89)
	at monix.reactive.internal.operators.SwitchMapObservable$$anon$1.onError(SwitchMapObservable.scala:96)
	at monix.reactive.internal.operators.SwitchMapObservable$$anon$1$$anon$2.onError(SwitchMapObservable.scala:74)
	at monix.reactive.observers.Subscriber$Implementation.onError(Subscriber.scala:214)
	at monix.reactive.observables.RefCountObservable$$anon$1.onError(RefCountObservable.scala:79)
	at monix.reactive.subjects.AsyncSubject.onCompleteOrError(AsyncSubject.scala:108)
	at monix.reactive.subjects.AsyncSubject.onError(AsyncSubject.scala:56)
	at monix.reactive.observers.Subscriber$Implementation.onError(Subscriber.scala:214)
	at monix.reactive.observers.SafeSubscriber.signalError(SafeSubscriber.scala:92)
	at monix.reactive.observers.SafeSubscriber.onError(SafeSubscriber.scala:61)
	at monix.reactive.internal.operators.DoOnErrorOperator$$anon$1.onError(DoOnErrorOperator.scala:46)
	at monix.reactive.internal.operators.DoOnNextOperator$$anon$1.onError(DoOnNextOperator.scala:55)
	at monix.reactive.internal.operators.MapTaskObservable$MapAsyncSubscriber.$anonfun$onNext$2(MapTaskObservable.scala:154)
	at monix.eval.Coeval$Always.runAttempt(Coeval.scala:603)
	at monix.eval.Task$.loop$1(Task.scala:1474)
	at monix.eval.Task$.startTrampolineForFuture(Task.scala:1519)
	at monix.eval.Task.runAsync(Task.scala:113)
	at monix.reactive.internal.operators.MapTaskObservable$MapAsyncSubscriber.onNext(MapTaskObservable.scala:179)
	at monix.reactive.internal.operators.DoOnTerminateOperator$$anon$1.onNext(DoOnTerminateOperator.scala:38)
	at monix.reactive.internal.operators.WhileBusyDropEventsOperator$$anon$1.onNext(WhileBusyDropEventsOperator.scala:40)
	at monix.reactive.internal.operators.WhileBusyDropEventsOperator$$anon$1.onNext(WhileBusyDropEventsOperator.scala:30)
	at monix.reactive.internal.operators.MapOperator$$anon$1.onNext(MapOperator.scala:45)
	at monix.reactive.internal.builders.IntervalFixedDelayObservable$$anon$1.run(IntervalFixedDelayObservable.scala:58)
	at scala.concurrent.impl.ExecutionContextImpl$AdaptedForkJoinTask.exec(ExecutionContextImpl.scala:140)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
```

**Comment by alexandru** [comment]
@YulawOne you are right, thanks a lot for the report and sorry for the inconvenience!

Unfortunately `AsyncSubject` hasn't seen so much usage, because it is redundant with `Task`, `Future` and operators such as `Observable.lastL.runAsync` and `Observable.runAsyncGetLast`, which also cache that last result.

It will get fixed in the next minor version (2.2.2) which will get released this week.

**Comment by alexandru** [comment]
Fix is now on `master` branch, see https://github.com/monix/monix/commit/0868ceb9035fe5c426a915559eb9b48e57efaca0

**Comment by alexandru** [comment]
@YulawOne sorry for the wait, this was fixed in today's release 2.2.2
